### ISSUE TITLE
Terraform 0.11.3 to 0.11.5 file provisioner regression 

### ISSUE THREAD

**Issue by bernadinm** [hashicorp/terraform#17733]
### Terraform Version

```
Affected Terraform Version

terraform -v
Terraform v0.11.5
+ provider.aws v1.13.0
+ provider.external v1.0.0
+ provider.null v1.0.0
+ provider.template v1.0.0

Working Terraform Version

$ terraform -v
Terraform v0.11.3
+ provider.aws v1.13.0
+ provider.external v1.0.0
+ provider.null v1.0.0
+ provider.template v1.0.0

Your version of Terraform is out of date! The latest version
is 0.11.5. You can update by downloading from www.terraform.io/downloads.html
```

### Terraform Configuration Files

```hcl
variable "aws_profile" { default  = "default"}
variable "aws_region" { default   = "us-west-2"}
variable "admin_cidr" { default   = "0.0.0.0/0"}
variable "ssh_key_name" { default   = "default"}
variable "num_of_gpu_agents" {  default = "1"}
variable "aws_gpu_agent_instance_type" {  default = "t2.micro"}
variable "aws_gpu_ami" { default = "ami-9b5d97fb"} # AMI Zone US-WEST-2

# Specify the provider and access details
provider "aws" {
  profile = "${var.aws_profile}"
  region = "${var.aws_region}"
}

# Create a VPC to launch our instances into
resource "aws_vpc" "default" {
  cidr_block = "10.0.0.0/16"
  enable_dns_hostnames = "true"
}

# Create an internet gateway to give our subnet access to the outside world
resource "aws_internet_gateway" "default" {
  vpc_id = "${aws_vpc.default.id}"
}

# Grant the VPC internet access on its main route table
resource "aws_route" "internet_access" {
  route_table_id         = "${aws_vpc.default.main_route_table_id}"
  destination_cidr_block = "0.0.0.0/0"
  gateway_id             = "${aws_internet_gateway.default.id}"
}

# Create a subnet to launch public nodes into
resource "aws_subnet" "public" {
  vpc_id                  = "${aws_vpc.default.id}"
  cidr_block              = "10.0.0.0/22"
  map_public_ip_on_launch = true
}

# Create a subnet to launch slave private node into
resource "aws_subnet" "private" {
  vpc_id                  = "${aws_vpc.default.id}"
  cidr_block              = "10.0.4.0/22"
  map_public_ip_on_launch = true
}


# A security group for Admins to control access
resource "aws_security_group" "admin" {
  name        = "admin-security-group"
  description = "Administrators can manage their machines"
  vpc_id      = "${aws_vpc.default.id}"

  # SSH access from anywhere
  ingress {
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    cidr_blocks = ["${var.admin_cidr}"]
  }

  # outbound internet access
  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}


resource "aws_instance" "node" {
  connection {
    user = "centos"
  }

  count = "${var.num_of_gpu_agents}"
  instance_type = "${var.aws_gpu_agent_instance_type}"
  tags { Name =  "terraform-bug-tester" }
  ami = "${var.aws_gpu_ami}"      
  key_name = "${var.ssh_key_name}"
  vpc_security_group_ids = ["${aws_security_group.admin.id}"]
  subnet_id = "${aws_subnet.private.id}"

  # OS init script
  provisioner "file" {
   source = "test.sh"
   #content = "true" # Works with this line but not with source (line above run $(touch test.sh)
   destination = "/tmp/os-setup.sh"
   }
}

output "GPU Private Public IP Address" {
  value = ["${aws_instance.node.*.public_ip}"]
}

```

### Crash Output

Terraform 0.11.5 Error https://gist.github.com/bernadinm/082a034e324f9adcbb52f4753fcf4fdf

### Expected Behavior

Working with Terraform 0.11.3 https://gist.github.com/bernadinm/428de43464f7a8c371cf6b157fd8f457

Same exact command.

### Steps to Reproduce

Please run the full steps to reproduce the issue:
1. Install terraform 0.11.5
2. Copy HCL template above and $(touch test.sh)
3. `terraform init`
4. `terraform apply`

**Comment by jbardin** [comment]
Thanks for filing the issue @bernadinm. This is an interesting failure, probably related to the Centos image.

The change that is causing this was one that allows the provisioner to actually fail on an authentication error, where as previously it would retry indefinitely. This was requested a number of times, since in most cases a failed authentication is never going to resolve itself (most likely a config error), and users would rather not wait for the entire operation times out while the authentication continues to fail. 

I'm not sure exactly what is different in the Centos image, whether the ssh daemon gets reconfigured and restarted, or maybe the user data with the authorized key is added late in the boot process.  I'll see what we can do to work around this.

**Comment by apparentlymart** [comment]
In images where `cloud-init` is used there may be a delay before the SSH credentials provided by the cloud platform (e.g. the public key in EC2) are installed into the appropriate place in the filesystem. On an image using systemd the socket activation might cause `sshd` to start up before `cloud-init` runs, causing it to not be "ready" yet.

If this is the cause, we might need to retry for some amount of time to give the image a chance to "settle". :confounded:

**Comment by jbardin** [comment]
That's a good idea on socket activation being the underlying cause. I haven't looked at the init process on Centos in a long time, but it is obviously doing something different than the ubuntu images.

**Comment by bernadinm** [comment]
@jbardin @apparentlymart. It looks like this is reproducible for any image that is used, not specifically towards Centos or the `cloud-init` scripts. 

I have just used a default Amazon Linux AMI below with the same script above, and it produces the same error. I am sure any image will have this same problem. 

`Amazon Linux AMI 2017.09.1 (HVM), SSD Volume Type - ami-d874e0a0`

**Comment by bernadinm** [comment]
I have just tested a SUSE Linux AMI and I have confirmed the same result.

`SUSE Linux Enterprise Server 12 SP3 (HVM), SSD Volume Type - ami-6bc56f13`'

It will successfully SSH if I change the method from  `source` to `content`. For whatever reason, it just does not work with `source`. 

```HCL
  # OS init script
  provisioner "file" {
   source = "test.sh"
   #content = "true" # Works with this line but not with source (line above run $(touch test.sh)
   destination = "/tmp/os-setup.sh"
   }
```

I believe this is actually related to issue #17692 here that another member reported as well. The same version of the terraform-0.11.5 was used.

**Comment by bernadinm** [comment]
I just tested with `Ubuntu Server 16.04 LTS (HVM), SSD Volume Type - ami-4e79ed36`. 

It also doesn't work with Ubuntu as well. At least we know is not the images themselves.

**Comment by bernadinm** [comment]
Thank you @jbardin and @apparentlymart. I have just confirmed the fix in master (terraform-0.11.6).
